### ISSUE TITLE
Link documents on a n:n relationship after they've been created

### ISSUE THREAD

**Issue by gjuchault** [neumino/thinky#291]
Hi,

When you call :
```js
let doc1 = new Doc1({ name: 'foo' });
doc1.save().then(function () {
  let doc2 = new Doc2({ docs1 : [doc1] });
  doc2.saveAll();
});
```

With a n:n relationship between doc1 and doc2; it tries to save doc1 which is already saved so there is a duplicate primary key error.
But how to add existing document to a n:n relationships ?

**Comment by neumino** [comment]
This works fine for me:

```js
var thinky = require('./lib/thinky.js')();
var type = thinky.type;

var Doc1 = thinky.createModel('doc1', {
  id: String
});
var Doc2 = thinky.createModel('doc2', {
  id: String
});

Doc2.hasAndBelongsToMany(Doc1, 'docs1', 'id', 'id');

var doc1 = new Doc1({ name: 'foo' });
doc1.save().then(function () {
  var doc2 = new Doc2({ docs1 : [doc1] });
  return doc2.saveAll()
}).then(function() {
  console.log('done');
});
```

**Comment by gjuchault** [comment]
This example is working indeed.
I think the problem comes from the fact that I'm working through an API.
The first call responds doc1, then I use this variable to make another JSON request call.
So thinky can't see that's an instance of the model..
For now, I uses ids but that works only for belongsTo/hasOne relationships..

Do you have any idea ?

**Comment by neumino** [comment]
I don't see your code, so it's a little hard to help you... but you can probably just save by passing an array of ids instead of an array of documents.

**Comment by gjuchault** [comment]
If i send the ids, it says that it can't build a new instance of `Doc1` without an object.

Basically, that's just a REST API. I post a new Doc1, I get back its document (with the new id) :
```js
POST /doc1 { name : "foo" }
{ id: "xxxxx", name: "foo" }
```
and i keep the result in a variable `doc1`.

Then I post a new Doc2, with multiple Doc1 as they have a n:n relationship
```js
POST /doc2 { docs1 : [ doc1.id ] }
Cannot build a new instance of `Article` without an object
```

Or
```js
POST /doc2 { docs1 : [ doc1 ] }
Duplicate primary key ...
```

If the relation is 1:1 or 1:n I can do
```js
POST /doc2 { doc1Id : doc1.id }
{ id: 'xxx', doc1: { id: 'xxx', name: 'foo' }, doc1Id: 'xxx' }
```
as there is a leftKey.

But if the relation is n:n, I just can't send the thinky object (which has method like `isSaved()` and all that you may use with `saveAll()`) so I send the body.

So how can I append, given the primary key, a document to the n:n relationship ?
Is my only way to use thinky.r.table('Doc2_Doc1') or something like that ?

**Comment by gjuchault** [comment]
Some help please ?

**Comment by gjuchault** [comment]
Is there another way to push a relation ship by id other than using `r` ?

**Comment by neumino** [comment]
You can retrieve `doc1` from the database and inject it in `doc2` before calling `doc2.saveAll({docs1: true})`.

**Comment by gjuchault** [comment]
Fact is I may want to have multiple relationships between the same documents (use case : a promotion when you buy 3 things, with Promotion n:n Articles).
Does thinky use a reference on the variable to prevent this behaviour (that is not a classic one) ? Or should I use a new variable each time I want to save the document ?

**Comment by neumino** [comment]
`hasAndBelongsToMany` represents a `n-n` relation, so I'm not sure to understand your question.

On a side not, you can also do this:

```
var doc2 = new Doc(...);
doc2.docs1 = [ doc1.id ];
doc2.saveAll({docs1: true})
```

**Comment by gjuchault** [comment]
But I did try :
```
POST /doc2 { docs1 : [ doc1.id ] }
Cannot build a new instance of `Doc1` without an object
```
Which will call
```js
var doc2 = new Doc2({ docs1 : [ XXXXX ] });
doc2.saveAll();
```

Okay, I see that i'm missing the options in saveAll.

1. Is that the cause of the error ?
2. What will happen if I call `doc2.docs1 = [ doc1.id, doc1.id ];` (using your example to explain my last comment) ? Will it create the relation two times as expected, or will it uniquify ?

**Comment by neumino** [comment]
Note that I define the field `docs1` after creating the document.

**Comment by gjuchault** [comment]
> Note that I define the field docs1 after creating the document.
That is that much important ? :frowning:

**Comment by gjuchault** [comment]
And will it create two times the same relation ship if you do `doc2.docs1 = [ doc1.id, doc1.id ]` ?

**Comment by neumino** [comment]
I believe it would not, but you may want to try yourself to be sure.

**Comment by gjuchault** [comment]
Okay; here's what I've done so far :
* Extracting basing on Model._joins the data from req.body just before instanciating the model.
* Instanciate the model
* Restores the keys extracted before (to use the setter you made I think, well, see comments above)
* Calling saveAll with the object as parameter to make thinky understand which documents related I want to save.

But the result of the saveAll command is interesting :
```js
var doc2 = new Doc2({});
var doc1 = new Doc1({ foo: 'bar' });
docs2.docs1 = [ doc1.id ];
doc2.saveAll().then(function (result) {
    assert.equals('bar', result.docs1[0].foo); // False; result.docs1 still equals to [ doc1.id ]
});
```

Basically, thinky doesn't returns the created object, but the passed object.

Is this the wanted behaviour of thinky ?
  -> Yes : Is it possible to have an option such as `returnChanges` in rethinkdb
  -> No   : Should I open a new issue for that ?

**Comment by gjuchault** [comment]
Of course I could do (basing on the example above) :
```
var doc2 = new Doc2({});
var doc1 = new Doc1({ foo: 'bar' });
docs2.docs1 = [ doc1.id ];
doc2.saveAll({ docs1: true }).then(function (result) {
    Doc1.get(result.id).getJoin({ docs1: true }).then(function (result) {
        // Here I should get result.docs1[0].foo
    });
    assert.equals('bar', result.docs1[0].foo); // False; result.docs1 still equals to [ doc1.id ]
});
```

**Comment by neumino** [comment]
Yes, this is the intended behavior. This is because the `n-n` relations, the relations are saved in another table, so returning the whole joined document when we just save a relation requires an extra read on the table.

There could probably be an option though.
